src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Formatting/RemoteFormattingService.cs

```diff
@@ -0,0 +1,175 @@
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var sourceText = await context.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (!sourceText.TryGetAbsoluteIndex(linePosition, out var hostDocumentIndex))
+        {
+            return [];
+        }
+
+        if (!_formattingService.TryGetOnTypeFormattingTriggerKind(codeDocument, hostDocumentIndex, triggerCharacter, out var triggerCharacterKind))
+        {
+            return [];
+        }
+
+        TextEdit[] result;
+        if (triggerCharacterKind is RazorLanguageKind.Html)
+        {
+            var htmlEdits = htmlChanges.Select(sourceText.GetTextEdit).ToArray();
+            result = await _formattingService.GetHtmlOnTypeFormattingEditsAsync(context, htmlEdits, options, hostDocumentIndex, triggerCharacter[0], cancellationToken).ConfigureAwait(false);
+        }
+        else if (triggerCharacterKind is RazorLanguageKind.CSharp)
+        {
+            result = await _formattingService.GetCSharpOnTypeFormattingEditsAsync(context, options, hostDocumentIndex, triggerCharacter[0], cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            Assumed.Unreachable();
+            return [];
+        }
+
+        return result.SelectAsArray(sourceText.GetTextChange);
+    }
+
+    public ValueTask<Response> GetOnTypeFormattingTriggerKindAsync(
+        RazorPinnedSolutionInfoWrapper solutionInfo,
+        DocumentId documentId,
+        LinePosition linePosition,
+        string triggerCharacter,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => IsValidOnTypeFormattingTriggerAsync(context, linePosition, triggerCharacter, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<Response> IsValidOnTypeFormattingTriggerAsync(RemoteDocumentContext context, LinePosition linePosition, string triggerCharacter, CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var sourceText = await context.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (!sourceText.TryGetAbsoluteIndex(linePosition, out var hostDocumentIndex))
+        {
+            return Response.Invalid;
+        }
+
+        if (!_formattingService.TryGetOnTypeFormattingTriggerKind(codeDocument, hostDocumentIndex, triggerCharacter, out var triggerCharacterKind))
+        {
+            return Response.Invalid;
+        }
+
+        if (triggerCharacterKind is RazorLanguageKind.Html)
+        {
+            return Response.ValidHtml;
+        }
+
+        return Response.ValidCSharp;
```

Comment:
From a readability standpoint, it feels a little presumptive that non-HTML is valid C#, especially since the value `RazorLanguageKind.Razor` exists. I know that the implementation of `TryGetOnTypeFormattingTriggerKind` only ever returns true for HTML and C#, but that's not necessarily obvious to the reader here.